### PR TITLE
[ConstraintSystem] Diagnose conditional requirement failures via fixes

### PR DESCRIPTION
**Pull request by xedin**
Extend existing `RequirementFailure` functionality to support
conditional requirement failures. Such fixes are introduced
only if the parent type requirement has been matched successfully.

Resolves: rdar://problem/47871590

### CODE REVIEW

File: test/Generics/conditional_conformances.swift [hunk 1]
```diff
@@ -425,3 +391,14 @@ protocol P {
...
 }
+
+// rdar://problem/47871590
+
+extension BinaryInteger {
+  var foo: Self {
+    return self <= 1
+            ? 1
+            : (2...self).reduce(1, *)
+            // expected-error@-1 {{instance method 'reduce' requires that 'Self.Stride' conform to 'SignedInteger'}}
```
Comment (xedin) [inline test/Generics/conditional_conformances.swift]: This diagnostic has a note:

```
note: requirement from conditional conformance of 'ClosedRange<Self>' to 'Sequence'
extension ClosedRange : Sequence where Bound : Strideable, Bound.Stride : SignedInteger {
          ^
```

File: lib/Sema/CSDiagnostics.cpp [hunk 2]
```diff
@@ -88,19 +90,54 @@ Type RequirementFailure::getOwnerType() const {
 }
 
 const GenericContext *RequirementFailure::getGenericContext() const {
+  // For conditional requirements, this is easy because all
+  // required information is contained in the conformance
+  // reference.
+  if (isConditional()) {
+    auto *conformance = getConformanceRef().getConcrete();
+    auto *DC = conformance->getDeclContext();
```
Comment (slavapestov) [inline lib/Sema/CSDiagnostics.cpp]: We should think of a better way to express this. I think I've see this appear a few times in this file.

File: test/Constraints/array_literal.swift [hunk 3]
```diff
@@ -338,7 +339,7 @@ extension Int: P { }
 
 func testConditional(i: Int, s: String) {
   let _: PArray<Int> = [i, i, i]
-  let _: PArray<String> = [s, s, s] // expected-error{{cannot convert value of type '[String]' to specified type 'PArray<String>'}}
+  let _: PArray<String> = [s, s, s] // expected-error{{generic struct 'PArray' requires that 'String' conform to 'P'}}
```
Comment (slavapestov) [inline test/Constraints/array_literal.swift]: Cool!

**Review by slavapestov — APPROVED** [review]

File: lib/Sema/CSDiagnostics.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Actually I realized that I can simply do `getRequirementDC()` in this case, no need to fetch conformance explicitly, thanks!

Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Hm, after thinking about this - it genericContext() should be related to the affect decl here and not the requirement, so this if shouldn't even be needed...

Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Fixed!

Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: That also improved diagnostic for:
```swift
extension BinaryInteger {
  var factorial: Self {
    return self <= 1
                ? 1
                : (2...self).reduce(1, *)
  }
}
```

To:
```
error: referencing instance method 'reduce' on 'ClosedRange' requires that 'Self.Stride' conform to 'SignedInteger'
                  : (2...self).reduce(1, *)
                               ^
note: requirement from conditional conformance of 'ClosedRange<Self>' to 'Sequence'
extension ClosedRange : Sequence where Bound : Strideable, Bound.Stride : SignedInteger {
          ^
```

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by xedin** [comment]
@shahmishal Looks like this is something with the bot itself in the final phase.

**Comment by xedin** [comment]
@swift-ci please test

File: test/Generics/conditional_conformances.swift [hunk 4]
```diff
@@ -12,55 +12,36 @@ protocol P6: P2 {}
...
-// expected-note@-14{{candidate requires that the types 'Float' and 'Int' be equivalent (requirement specified as 'Float' == 'Int')}}
-// expected-note@-15{{requirement from conditional conformance of 'Infer<Constrained<U>, V>' to 'P2'}}
-// expected-note@-16{{candidate requires that the types 'Constrained<U>' and 'Constrained<V>' be equivalent (requirement specified as 'Constrained<U>' == 'Constrained<V>')}}
-// expected-note@-17{{candidate requires that the types 'U' and 'Int' be equivalent (requirement specified as 'U' == 'Int')}}
-// expected-note@-18{{requirement from conditional conformance of 'SameType<Float>' to 'P2'}}
-// expected-note@-19{{requirement from conditional conformance of 'SameType<U>' to 'P2'}}
+// expected-note@-1 {{candidate requires that the types 'U' and 'V' be equivalent (requirement specified as 'U' == 'V')}}
+// expected-note@-2 {{requirement from conditional conformance of 'SameTypeGeneric<U, V>' to 'P2'}}
+// expected-note@-3 {{requirement from conditional conformance of 'SubclassBad' to 'P2'}}
+// expected-note@-4 {{candidate requires that 'C1' inherit from 'U' (requirement specified as 'U' : 'C1')}}
```
Comment (harlanhaskins) [inline test/Generics/conditional_conformances.swift]: Should this read `candidate requires that 'U' inherit from 'C1'`?

Comment (xedin) [reply test/Generics/conditional_conformances.swift]: Yeah, probably, I didn't touch existing messages in declaration checker.

**Pull request merged**